File: backend/maze.py
# Author __Pham Ngoc Hai - Harito__
# ---------------------------------
import numpy as np

class Maze:
    # Create a maze with rows, cols and obstacle_prob para or create a Maze with random maze value
    def __init__(self, shape: tuple, obstacle_prob=0.3, maze=None):
        self.rows = shape[0]
        self.cols = shape[1]
        self.maze = maze
        if maze is None:
            self.__generate__(obstacle_prob)

    def __generate__(self, obstacle_prob):
        random_values = np.random.rand(self.rows, self.cols)
        self.maze = np.full((self.rows, self.cols), False, dtype=bool)
        self.maze[random_values < obstacle_prob] = True

    @classmethod
    def array(cls, array):
        maze = np.array(array, dtype=bool)
        return cls(maze.shape, maze=maze)

    # Change maze follow by probability and check the updated maze can have path to go or not
    def change_maze(self, prob_false_to_true, prob_true_to_false):
        is_dead_maze = 0
        for i in range(self.rows):
            for j in range(self.cols):
                random_value = np.random.rand()
                if not self.maze[i, j]:
                    if random_value < prob_false_to_true:
                        self.maze[i, j] = True
                    else:
                        is_dead_maze += 1
                elif self.maze[i, j] and random_value < prob_true_to_false:
                    self.maze[i, j] = False
                    is_dead_maze += 1
        return is_dead_maze == 0

    # Return a clone maze (type: Maze)
    def copy(self):
        return Maze((self.rows, self.cols), maze=np.copy(self.maze))

    # Check the current position can go throw or not
    def is_valid(self, current: tuple):
        return current[0] >= 0 and current[1] >= 0 \
            and current[0] < self.rows \
            and current[1] < self.cols \
            and not self.maze[current[0], current[1]]

    # Return a list of neighbors can reach of the current position
    def get_neighbors(self, current: tuple):
        neighbors = []
        directions = ((-1, 0), (0, 1), (1, 0), (0, -1))
        for move in directions:
            neighbor = (current[0] + move[0], current[1] + move[1])
            if self.is_valid(neighbor):
                neighbors.append(neighbor)
        return neighbors

    # Return the String describe maze
    def __str__(self):
        return self.maze.__str__()
    
    # Print maze
    def print(self):
        print(self.__str__())

    # # Solve 
    # def solve(self, method: str, type: str):
    #     if method == 'AStar':
    #         if type == 'Path':
    #             return AStar.path(self.maze)
    #         elif type == 'Map':
    #             return AStar.map(self.maze)
    #         else:
    #             return Exception('Only support "Path" and "Map"')
    #     elif method == 'ACO':
    #         return 
    #     else:
    #         return Exception('Only support "AStar" and "ACO"')


def main():
    # Sử dụng lớp Maze
    rows = 5  # Số hàng
    cols = 5  # Số cột
    obstacle_prob = 0.3  # Xác suất ô bị chặn

    maze_instance = Maze((rows, cols), obstacle_prob)

    # Gọi các phương thức của lớp Maze
    print("Initial Maze with random value:")
    maze_instance.print()

    print("Maze with selected value: ")
    maze2 = Maze.array([[True, True, False],
                        [True, True, False],
                        [True, True, False],
                        [True, True, False],
                        [True, True, False]])
    maze2.print()

    is_dead_maze = maze_instance.change_maze(0.5, 0)
    print("Changed the Random Maze:")
    maze_instance.print()
    print("Is Dead Maze:", is_dead_maze)

    cloned_maze = maze_instance.copy()
    print("Cloned the Random Maze:")
    cloned_maze.print()

    is_dead_maze = cloned_maze.change_maze(0.2, 0)
    print("Change cloned Maze:")
    cloned_maze.print()
    print("Is Dead Maze:", is_dead_maze)

if __name__ == "__main__":
    main()

File: backend/Maze.cpp
#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>

class Maze
{
private:
    int rows;
    int cols;
    std::vector<std::vector<bool>> maze;

    void generate(double obstacle_prob)
    {
        maze.resize(rows, std::vector<bool>(cols, false));

        // Seed bộ sinh số ngẫu nhiên dựa trên thời gian
        std::srand(static_cast<unsigned>(std::time(nullptr)));

        for (int i = 0; i < rows; ++i)
        {
            for (int j = 0; j < cols; ++j)
            {
                double random_value = static_cast<double>(std::rand()) / RAND_MAX;
                if (random_value < obstacle_prob)
                {
                    maze[i][j] = true;
                }
            }
        }
    }

public:
    Maze(std::pair<int, int> shape, double obstacle_prob = 0.3)
    {
        rows = shape.first;
        cols = shape.second;
        generate(obstacle_prob);
    }

    Maze(std::initializer_list<std::vector<bool>> values)
    {
        for (std::vector<bool> row : values)
        {
            maze.push_back(row);
        }
        rows = maze.size();
        cols = maze.at(0).size();
    }

    Maze(std::vector<std::vector<bool>> maze)
    {
        this -> maze = maze;
        rows = maze.size();
        cols = maze.at(0).size();
    }

    int getHeight()
    {
        return rows;
    }

    int getWidth()
    {
        return cols;
    }

    // Hàm chọn ngẫu nhiên hai ô có giá trị false làm điểm bắt đầu và điểm kết thúc
    // Chọn điểm bắt đầu và kết thúc sau khi tạo maze mà không phải trước đó
    // vì để bám sát các vấn đề thực tiễn ứng dụng là map có trước
    std::pair<std::pair<int, int>, std::pair<int, int>> selectStartAndEnd()
    {
        int m = maze.size();
        int n = maze[0].size();

        std::pair<int, int> start, end;

        // Lặp để chọn ngẫu nhiên hai ô false
        while (true)
        {
            int x1 = std::rand() % m;
            int y1 = std::rand() % n;
            int x2 = std::rand() % m;
            int y2 = std::rand() % n;

            if (!maze[x1][y1] && !maze[x2][y2] && (x1 != x2 || y1 != y2))
            {
                start = std::make_pair(x1, y1);
                end = std::make_pair(x2, y2);
                break;
            }
        }

        return std::make_pair(start, end);
    }

    bool changeMaze(double prob_false_to_true, double prob_true_to_false)
    {
        int is_dead_maze = 0;
        for (int i = 0; i < rows; ++i)
        {
            for (int j = 0; j < cols; ++j)
            {
                double random_value = (double)rand() / RAND_MAX;
                if (!maze[i][j])
                {
                    if (random_value < prob_false_to_true)
                    {
                        maze[i][j] = true;
                    }
                    else
                    {
                        is_dead_maze += 1;
                    }
                }
                else if (maze[i][j] && random_value < prob_true_to_false)
                {
                    maze[i][j] = false;
                    is_dead_maze += 1;
                }
            }
        }
        return is_dead_maze == 0;
    }

    Maze copy()
    {
        return Maze(maze);
    }

    bool isValid(std::pair<int, int> current)
    {
        return current.first >= 0 && current.second >= 0 &&
               current.first < rows && current.second < cols &&
               !maze[current.first][current.second];
    }

    std::vector<std::pair<int, int>> getNeighbors(std::pair<int, int> current)
    {
        std::vector<std::pair<int, int>> neighbors;
        std::vector<std::pair<int, int>> directions = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};
        for (const auto &move : directions)
        {
            std::pair<int, int> neighbor = {current.first + move.first, current.second + move.second};
            if (isValid(neighbor))
            {
                neighbors.push_back(neighbor);
            }
        }
        return neighbors;
    }

    void print()
    {
        for (const auto &row : maze)
        {
            for (bool cell : row)
            {
                std::cout << (cell ? "1 " : "0 ");
            }
            std::cout << std::endl;
        }
    }
};

// int main()
// {
//     Maze myMaze({5, 5});
//     // Maze myMaze = {{false, false, true}, {false, true, false}, {false, false, false}};
//     std::cout << "The origen maze..." << std::endl;
//     myMaze.print();
    
//     std::cout << "The copy maze: " << std::endl;
//     Maze copyMaze = myMaze.copy();
//     copyMaze.print();

//     std::cout << "Changing origen maze..." << std::endl;
//     myMaze.changeMaze(0.5, 0.0);
//     myMaze.print();

//     std::cout << "The copy maze: " << std::endl;
//     copyMaze.print();
//     return 0;
// }
